tensorflow/core/kernels/dml_kernel_wrapper.cc

```diff
@@ -103,6 +103,22 @@ void DmlKernelWrapperBase::Compute(OpKernelContext* ctx) {
       return;
     }
 
+    int inputIndexToForward = -1;
+    if (shared_helper->IsInputForwardable(ctx, output_shapes,
+                                          inputIndexToForward)) {
+      // This should be set if true was returned
+      CHECK(inputIndexToForward != -1);
+      // We assume we are forwarding one input to one output for now.
+      CHECK(ctx->num_outputs() == 1);
+      constexpr int outputIndex = 0;
+      Tensor* output;
+      CHECK(ctx->forward_input_to_output_with_shape(
+          inputIndexToForward, outputIndex, output_shapes[outputIndex],
+          &output));
```

Comment:
Is there a reason we do this instead of the more commonly used `set_output` found in most CPU/CUDA kernels? I think using `forward_input_to_output_with_shape` is risky because it can fail for legitimate reasons, whereas `set_output` should only fail if there's not enough memory (which is why it returns void). The kernel has parameters that tell us whether an input can be forwarded to an output or not and `set_output` handles all that heavy lifting for us.

Reply:
The idea is to essentially eliminate the cost of identity reductions by forwarding the tensors rather than dispatching or copying. `set_output` performs a copy.

However, it's a good point that we should not simply assert success with `CHECK` because the call to `forward_input_to_output_with_shape` can fail, like Pat says. We should attempt to forward but fall back on failure to doing a copy.

Reply:
That's exactly my point though: `set_output` handles all of that for us. It first tries to forward, but if there's a `Params::kNeverForward` flag, it will fall back to allocating and copying. We don't need to bother with the fallback ourselves.

Reply:
You're right. I was just reading the stale comments in the header. :)